### PR TITLE
Wallet cache

### PR DESCRIPTION
**Pull request by Sajma27**
# Summary
A description of the changes proposed in the pull request.

Fixes #number

## Checklist
- [x] Change is small and easy to review (please split big changes into multiple PRs)
- [x] Change is consistent with architecture
- [x] Change is consistent with test architecture
- [x] Change is consistent with naming conventions
- [x] New code is covered with tests
- [ ] Tests related to old code are updated
- [ ] Documentation is up to date with changes

### CODE REVIEW

File: universal-login-relayer/test/integration/ethereum/WalletContractService.test.ts [hunk 1]
```diff
@@ -36,6 +41,47 @@ describe('INT: WalletContractService', () => {
...
+      const version = await walletContractService.getWalletVersion(proxyContract.address);
+      expect(version).to.eq('beta2');
+      expect(fetchWalletVersionSpy).to.be.calledOnce;
+    });
+
+    it('adding cache after calling address with empty cache', async () => {
+      await walletContractService.getWalletVersion(proxyContract.address);
+      expect((walletContractService as any).walletVersions).to.deep.eq({[proxyContract.address]: 'beta2'});
+      expect(fetchWalletVersionSpy).to.be.calledOnce;
+    });
```
Comment (JustynaBroniszewska) [inline universal-login-relayer/test/integration/ethereum/WalletContractService.test.ts]: You can merge these two in one test `returns proper wallet version and saves it to the cache`

File: universal-login-relayer/src/integration/ethereum/WalletContractService.ts [hunk 2]
```diff
@@ -22,6 +25,21 @@ export class WalletContractService {
...
+      return this.walletVersions[walletAddress];
+    } else {
+      ensure(walletAddress !== undefined && isProperAddress(walletAddress), Error, 'Invalid address provided');
+      const walletVersion = await this.blockchainService.fetchWalletVersion(walletAddress);
+      this.walletVersions[walletAddress] = walletVersion;
+      return walletVersion;
+    }
+  }
+
+  async clearCatch() {
```
Comment (Marik-D) [inline universal-login-relayer/src/integration/ethereum/WalletContractService.ts]: Rename to `clearCache`

File: universal-login-relayer/test/integration/ethereum/WalletContractService.test.ts [hunk 3]
```diff
@@ -36,6 +41,47 @@ describe('INT: WalletContractService', () => {
...
+
+    it('address with cache', async () => {
```
Comment (JustynaBroniszewska) [inline universal-login-relayer/test/integration/ethereum/WalletContractService.test.ts]: `restores wallet version from cache`

File: universal-login-relayer/test/integration/ethereum/WalletContractService.test.ts [hunk 4]
```diff
@@ -36,6 +41,47 @@ describe('INT: WalletContractService', () => {
...
+      await walletContractService.getWalletVersion(proxyContract.address);
+      expect((walletContractService as any).walletVersions).to.deep.eq({[proxyContract.address]: 'beta2'});
+      await walletContractService.getWalletVersion(proxyContract.address);
+      expect(fetchWalletVersionSpy).to.be.calledOnce;
+    });
+
+    it('random address', async () => {
```
Comment (JustynaBroniszewska) [inline universal-login-relayer/test/integration/ethereum/WalletContractService.test.ts]: throws an error if a random address provided

File: universal-login-relayer/src/integration/ethereum/WalletContractService.ts (hunk 2 above)
Comment (JustynaBroniszewska) [inline universal-login-relayer/src/integration/ethereum/WalletContractService.ts]: I am not sure if we need this function somewhere outside the tests

File: universal-login-relayer/test/integration/ethereum/WalletContractService.test.ts [hunk 5]
```diff
@@ -194,4 +240,9 @@ describe('INT: WalletContractService', () => {
       expect(await walletContractService.isValidMessageHash(proxyContract.address, '0x1234', {} as any)).to.be.true;
     });
   });
+
+  afterEach(() => {
```
Comment (JustynaBroniszewska) [inline universal-login-relayer/test/integration/ethereum/WalletContractService.test.ts]:
```ts
afterEach(() => {
    fetchWalletVersionSpy.resetHistory();
    (walletContractService as any).walletVersions = {};
  });
```

File: universal-login-relayer/src/integration/ethereum/WalletContractService.ts [hunk 6]
```diff
@@ -1,16 +1,19 @@
...
+  private walletVersions: Record<string, WalletVersion | undefined> = {};
+
+  constructor(private blockchainService: BlockchainService, private beta2Service: Beta2Service, private gnosisSafeService: GnosisSafeService) {
 
   }
 
   private async getWalletService(walletAddress: string): Promise<IWalletContractService> {
-    const walletVersion = await this.blockchainSerivce.fetchWalletVersion(walletAddress);
+    const walletVersion = await this.getWalletVersion(walletAddress);
+
```
Comment (JustynaBroniszewska) [inline universal-login-relayer/src/integration/ethereum/WalletContractService.ts]: We don't need a new line here

File: universal-login-relayer/test/integration/ethereum/WalletContractService.test.ts [hunk 7]
```diff
@@ -36,6 +41,37 @@ describe('INT: WalletContractService', () => {
...
+  describe('version cache', () => {
+    it('throws an error if address is undefined', async () => {
+      await expect(walletContractService.getWalletVersion(undefined as any)).to.be.rejectedWith('Invalid address provided');
+    });
+
+    it('throws an error if address is empty', async () => {
+      await expect(walletContractService.getWalletVersion('')).to.be.rejectedWith('Invalid address provided');
+    });
+
+    it('returns proper wallet version and saves it to the cache', async () => {
```
Comment (JustynaBroniszewska) [inline universal-login-relayer/test/integration/ethereum/WalletContractService.test.ts]: I would change the order of operations: 
- first, we call getWalletVersion, 
- then we check if it returns the proper version and stores it in walletVersion
- then we can call it again and check if it was called once

```ts
it('returns proper wallet version and saves it to the cache', async () => {
      const version = await walletContractService.getWalletVersion(proxyContract.address);
      expect(version).to.eq('beta2');
      expect((walletContractService as any).walletVersions).to.deep.eq({[proxyContract.address]: 'beta2'});
      await walletContractService.getWalletVersion(proxyContract.address);
      expect(fetchWalletVersionSpy).to.be.calledOnce;
    });
```

**Pull request merged**